### PR TITLE
Map rotation rebase

### PR DESCRIPTION
**Pull request by DjWarmonger**
Map coordinates were reordered so now the map has shape of [z][x][y]. This has numerous advantages:

- For the most part, we are interested in only surface or underground. Now the tiles on same plane are located closer in memory, which should improve cache access both for pathfinding and rendering.
- The innermost loop iterating over 1 or 2 indexes is a waste of CPU cycles.
- Some abysmal `vector<vector<vector>>>` was replaced by `multi_array`, too.
- Thanks to `boost::multi_array` we can slice the interesting plane of the map. This already has its use in renderer and will probably be useful in the future, with editor or GUI rewrite.

Additionally, Nullkiller AI got similiar treatment. Now the outermost dimension is layer (ground, fly, waterwalk) which rarely changes and is often unneccessary. The innermost dimension are chain nodes which are processed together in the innermost loop.

Also, I'm shocked about apparent RMG speedup compared to 1.0.0 I just installed today. XL Jebus, no U, 4 players, normal difficulty.

<html>
<head>
	

<body>

  | 1.0.0 | Map rotation
-- | -- | --
  | 7771 | 5392
  | 7369 | 5527
  | 7627 | 5611
  | 8278 | 5632
  | 7785 | 5638
  | 7239 | 5642
  | 8042 | 5689
  | 7917 | 5218
  | 7734 | 5471
  | 7556 | 5704
mean time (ms) | 7732 | 5552
Speed |   | 71,81%
Improvement |   | 28,19%

</body>

</html>

### CODE REVIEW

**Comment by Nordsoft91** [comment]
1. So many changes related to spaces. I don't see problem with that in general, but It can create very annoying merge conflicts. Is it possible to omit changes in tabs/new lines to reduce amount of files impacted?
2. Please, rebase over last develop, some of the diffs show changes already being in develop

**Comment by DjWarmonger** [comment]
1. Actually, there were some spaces in empty lines which causes conflicts for me as well.

**Comment by DjWarmonger** [comment]
2. I don't think so, everything but formatting looks fine to me.

**Comment by Nordsoft91** [comment]
How it's possible? I mean, those changes are already in develop branch so shouldn't create any merge conflicts if your branch begins from recent commits

**Comment by Nordsoft91** [comment]
Terrains is quite new file which nobody touched yet. Conflicts are just impossible unless you have another branch which you use as base

**Comment by DjWarmonger** [comment]
No idea, I spent more than a hour cleaning up this mess already.

File: lib/Terrain.h [hunk 1]
```diff
@@ -18,16 +18,16 @@
 class DLL_LINKAGE Terrain
 {
 public:
-	
+
```
Comment (Nordsoft91) [inline lib/Terrain.h:21]: Changes here are useless, just discard this file from commit

File: lib/rmg/ObjectManager.h [hunk 2]
```diff
@@ -39,29 +39,29 @@ class ObjectManager: public Modificator
 	
 public:
 	MODIFICATOR(ObjectManager);
-	
+
```
Comment (Nordsoft91) [inline lib/rmg/ObjectManager.h:42]: This file shouldn't be a part of review too

File: AI/VCAI/AIUtility.h [hunk 3]
```diff
@@ -154,10 +154,62 @@ struct creInfo
 };
 creInfo infoFromDC(const dwellingContent & dc);
 
-void foreach_tile_pos(std::function<void(const int3 & pos)> foo);
-void foreach_tile_pos(CCallback * cbp, std::function<void(CCallback * cbp, const int3 & pos)> foo); // avoid costly retrieval of thread-specific pointer
-void foreach_neighbour(const int3 & pos, std::function<void(const int3 & pos)> foo);
-void foreach_neighbour(CCallback * cbp, const int3 & pos, std::function<void(CCallback * cbp, const int3 & pos)> foo); // avoid costly retrieval of thread-specific pointer
+template<class Func>
+void foreach_tile_pos(const Func & foo)
```
[the remark this answers is not in the document]
Comment (DjWarmonger) [reply AI/VCAI/AIUtility.h:158]: Nonono, they were in **Nullkiller AI**. I copied them straight into **VCAI** as well.

We can consider some common include file, though.

File: AI/VCAI/AIUtility.cpp [hunk 4]
```diff
@@ -137,55 +137,6 @@ bool HeroPtr::operator==(const HeroPtr & rhs) const
 	return h == rhs.get(true);
 }
 
-void foreach_tile_pos(std::function<void(const int3 & pos)> foo)
```
Comment (Nordsoft91) [inline AI/VCAI/AIUtility.cpp:140]: Those changes shouldn't be a part of review

Comment (Nordsoft91) [inline AI/VCAI/AIUtility.cpp:140]: oh, got it! same change as it was done for kai, but for vcmi

File: lib/Terrain.cpp [hunk 5]
```diff
@@ -256,4 +256,4 @@ bool Terrain::isNative() const
 bool Terrain::isTransitionRequired() const
 {
 	return Terrain::Manager::getInfo(*this).transitionRequired;
-}
```
Comment (Nordsoft91) [inline lib/Terrain.cpp:259]: Consider discarding changes from this file too

File: lib/mapping/CMapEditManager.h [hunk 6]
```diff
@@ -375,6 +375,7 @@ class CDrawTerrainOperation : public CMapOperation
 	InvalidTiles getInvalidTiles(const int3 & centerPos) const;
 
 	void updateTerrainViews();
+
```
Comment (Nordsoft91) [inline lib/mapping/CMapEditManager.h:378]: that's single change in the whole file

File: lib/Terrain.h [hunk 7]
```diff
@@ -1,12 +1,12 @@
 /*
- * Terrain.h, part of VCMI engine
- *
- * Authors: listed in file AUTHORS in main folder
- *
- * License: GNU General Public License v2.0 or later
- * Full text of license available in license.txt file, in main folder
- *
- */
+* Terrain.h, part of VCMI engine
```
Comment (Nordsoft91) [inline lib/Terrain.h:2]: Disagree with that, check other files - there are single spaces everywhere in header

File: client/mapHandler.cpp [hunk 8]
```diff
@@ -52,20 +52,22 @@ struct NeighborTilesInfo
...
 	{
 		auto getTile = [&](int dx, int dy)->bool
 		{
 			if ( dx + pos.x < 0 || dx + pos.x >= sizes.x
 			  || dy + pos.y < 0 || dy + pos.y >= sizes.y)
 				return false;
-			return settings["session"]["spectate"].Bool() ? true : visibilityMap[dx+pos.x][dy+pos.y][pos.z];
+
+			//FIXME: please do not read settings for every tile...
+			return settings["session"]["spectate"].Bool() ? true : (*visibilityMap)[pos.z][dx+pos.x][dy+pos.y];
```
Comment (Nordsoft91) [inline client/mapHandler.cpp:64]: is it possible to make it static and capture by lambda?

File: client/mapHandler.cpp [hunk 9]
```diff
@@ -557,7 +544,9 @@ void CMapHandler::CMapWorldViewBlitter::drawTileOverlay(SDL_Surface * targetSurf
 		const CGObjectInstance * obj = object.obj;
 
 		const bool sameLevel = obj->pos.z == pos.z;
-		const bool isVisible = settings["session"]["spectate"].Bool() ? true : (*info->visibilityMap)[pos.x][pos.y][pos.z];
+
+		//FIXME: Don't read options in  a loop :v
+		const bool isVisible = settings["session"]["spectate"].Bool() ? true : (*info->visibilityMap)[pos.z][pos.x][pos.y];
```
Comment (Nordsoft91) [inline client/mapHandler.cpp:549]: Looks like easy to fix and make single check out of cycle

File: lib/Terrain.cpp [hunk 10]
```diff
@@ -46,7 +46,7 @@ Terrain::Manager::Manager()
 	{
 		if(!CResourceHandler::get(mod)->existsResource(ResourceID("config/terrains.json")))
 			continue;
-		
+
```
Comment (Nordsoft91) [inline lib/Terrain.cpp:49]: All changes here looks also unnecessary

**Pull request closed without merging**